### PR TITLE
DeepSpeedZeroOptimizer dist.reduce optimization when self.round_robin_gradients is false

### PR DESCRIPTION
**Pull request by Liangliang-Ma**
This is an optimization for reduce behavior in ZeRO stage 2.

ZeRO stage 2 assigns ranks to trainable parameters at initial stage. It distributes parameters of each parameter group evenly by total mem size of the group. And it splits the param into two ranks when the param is across the division line. These two factors will cause adjacent parameters get sent to far apart ranks and show interlacing of ranks. When model doing backwards, we can observe a fragmentation of ranks in reduce operations. Each reduce has a small tensor size and the total number of reduce is big, which will result in a low performance.

Due to reduce bucket has a feature of combining tensors together if they are adjacent and share a common rank, we make an optimization to adjust the partition allocation of each parameter in order to maximize the combining chance of reduce operations. Basically we list all parameters of all groups in the order of backwards path and set the partition increasingly, instead of allocating parameters evenly in each groups. Also, this optimization will not change the parameters' group and any other attributes.

Only the partition is adjusted for each parameter and its grad. This will make the rank number of each reduce call in backwards shows a monotonically increasing and decrease the number of total reduce operation after combining in buckets. And finally it will improve the performance. In our workload we can observe five times less of number of reduce operations and a significant time shorten.

This feature will be turned on when round_robin_gradients flag is false.

### CODE REVIEW

**Comment by tjruwase** [comment]
@Liangliang-Ma, thanks for this PR and explanation. This optimization sounds really exciting. I will start reviewing asap. 

In the meantime, I have a few questions:
1. What model sizes does this optimization benefit most?
2. How this optimization affected by reduce_bucket_size configuration?
3. Did you test this for zero-offload?
4. How does the perf compare with `round_robin_gradients`?

**Comment by Liangliang-Ma** [comment]
Thanks for reviewing, @tjruwase !

> * What model sizes does this optimization benefit most?
This optimization is for zero stage 2. So not very big size model will get benefit. Currently we do experiments on 3.6 billion parameters and have good improvements using this.

> * How this optimization affected by reduce_bucket_size configuration?
reduce_bucket_size does not have too much affection unless it is too small.

> * Did you test this for zero-offload?
Not before. I tested it and found some issue.  A flag is added to exclude zero-offload. I think I can provide support for zero-offload by next PR :)
btw MoE will not run with this optimization because its world_sizes are different across param_groups.

> * How does the perf compare with `round_robin_gradients`?
Total reduce time decrease by 56% and reduce operation number decrease by 82% (12 ranks) in our workload.

**Comment by Liangliang-Ma** [comment]
Additionally, I attach the log of `rank_and_offset` before and after optimization. You can see the ranks of each reduce optimization are no longer messy.

before:
`[2022-12-08 08:16:34,989] [INFO] [stage_1_and_2.py:916:print_rank_0] [(11, 0, 264330240), (10, 264330240, 1024.0), (11, 264331264.0, 11264.0), (10, 264342528.0, 5472256.0), (11, 269814784.0, 32276480.0), (10, 302091264.0, 188823552)]
[2022-12-08 08:16:36,202] [INFO] [stage_1_and_2.py:916:print_rank_0] [(10, 0, 75515904), (9, 75515904, 5120.0), (10, 75521024.0, 1024.0), (9, 75522048.0, 10947584.0), (10, 86469632.0, 26804224.0), (9, 113273856.0, 264333312), (8, 377607168.0, 9216), (9, 377616384.0, 9437184), (8, 387053568.0, 16425984.0), (9, 403479552.0, 11894784.0), (8, 415374336.0, 75528192)]
[2022-12-08 08:16:37,010] [INFO] [stage_1_and_2.py:916:print_rank_0] [(8, 0, 151044096), (7, 151044096, 4096.0), (8, 151048192.0, 2048.0), (7, 151050240.0, 3072), (8, 151053312.0, 37748736), (7, 188802048.0, 21901312.0), (8, 210703360.0, 15859712.0), (7, 226563072.0, 226572288), (6, 453135360.0, 5120.0), (7, 453140480.0, 1024.0), (6, 453141504.0, 3072), (7, 453144576.0, 9437184), (6, 462581760.0, 9216), (7, 462590976.0, 28311552), (6, 490902528.0, 9216)]
[2022-12-08 08:16:38,153] [INFO] [stage_1_and_2.py:916:print_rank_0] [(6, 0, 27361280.0), (7, 27361280.0, 10387456.0), (6, 37748736.0, 188814336), (5, 226563072.0, 3072), (6, 226566144.0, 3072), (5, 226569216.0, 3072), (6, 226572288.0, 37748736), (5, 264321024.0, 12288), (6, 264333312.0, 37748736), (5, 302082048.0, 4531200.0), (6, 306613248.0, 4915200.0), (5, 311528448.0, 179377152)]
[2022-12-08 08:16:38,971] [INFO] [stage_1_and_2.py:916:print_rank_0] [(5, 0, 37748736), (4, 37748736, 4096.0), (5, 37752832.0, 2048.0), (4, 37754880.0, 3072), (5, 37757952.0, 9437184), (4, 47195136.0, 9216), (5, 47204352.0, 28311552), (4, 75515904.0, 566272.0), (5, 76082176.0, 37191680.0), (4, 113273856.0, 188814336), (3, 302088192.0, 2048.0), (4, 302090240.0, 4096.0), (3, 302094336.0, 3072), (4, 302097408.0, 37748736), (3, 339846144.0, 12288), (4, 339858432.0, 37748736), (3, 377607168.0, 9216), (4, 377616384.0, 9437184), (3, 387053568.0, 6038528.0), (4, 393092096.0, 22282240.0), (3, 415374336.0, 75528192)]
[2022-12-08 08:16:39,781] [INFO] [stage_1_and_2.py:916:print_rank_0] [(3, 0, 113276928), (2, 113276928, 3072), (3, 113280000, 3072), (2, 113283072, 3072), (3, 113286144, 9437184), (2, 122723328, 9216), (3, 122732544, 28311552), (2, 151044096, 9216), (3, 151053312, 37748736), (2, 188802048, 11513856.0), (3, 200315904.0, 26247168.0), (2, 226563072.0, 151053312), (1, 377616384.0, 1024.0), (2, 377617408.0, 5120.0), (1, 377622528.0, 3072), (2, 377625600.0, 37748736), (1, 415374336.0, 12288), (2, 415386624.0, 37748736), (1, 453135360.0, 9216), (2, 453144576.0, 9437184), (1, 462581760.0, 9216), (2, 462590976.0, 28311552), (1, 490902528.0, 9216)]
[2022-12-08 08:16:41,350] [INFO] [stage_1_and_2.py:916:print_rank_0] [(1, 0, 16973824.0), (2, 16973824.0, 20774912.0), (1, 37748736.0, 151047168), (0, 188795904.0, 2048.0), (1, 188797952.0, 4096.0), (0, 188802048.0, 3072), (1, 188805120.0, 9437184), (0, 198242304.0, 9216), (1, 198251520.0, 28311552), (0, 226563072.0, 9216), (1, 226572288.0, 37748736), (0, 264321024.0, 12288), (1, 264333312.0, 37748736), (0, 302082048.0, 9216), (1, 302091264.0, 9437184), (0, 311528448.0, 22455296.0), (1, 333983744.0, 5865472.0), (0, 339849216.0, 119583744)]
[2022-12-08 08:16:41,569] [INFO] [stage_1_and_2.py:916:print_rank_0] [(0, 0, 154533888)]`

after:
`[2022-12-08 05:59:05,014] [INFO] [stage_1_and_2.py:994:print_rank_0] [(0, 0, 296617984.0), (1, 296617984.0, 194296832.0)]
[2022-12-08 05:59:05,926] [INFO] [stage_1_and_2.py:994:print_rank_0] [(1, 0, 102321152.0), (2, 102321152.0, 296617984.0), (3, 398939136.0, 91963392.0)]
[2022-12-08 05:59:06,646] [INFO] [stage_1_and_2.py:994:print_rank_0] [(3, 0, 204654592.0), (4, 204654592.0, 286257152.0)]
[2022-12-08 05:59:07,497] [INFO] [stage_1_and_2.py:994:print_rank_0] [(4, 0, 10360832.0), (5, 10360832.0, 296617984.0), (6, 306978816.0, 183926784.0)]
[2022-12-08 05:59:08,230] [INFO] [stage_1_and_2.py:994:print_rank_0] [(6, 0, 112691200.0), (7, 112691200.0, 296617984.0), (8, 409309184.0, 81593344.0)]
[2022-12-08 05:59:08,954] [INFO] [stage_1_and_2.py:994:print_rank_0] [(8, 0, 215024640.0), (9, 215024640.0, 275887104.0)]
[2022-12-08 05:59:10,381] [INFO] [stage_1_and_2.py:994:print_rank_0] [(9, 0, 20730880.0), (10, 20730880.0, 296617984.0), (11, 317348864.0, 142084096.0)]
[2022-12-08 05:59:10,441] [INFO] [stage_1_and_2.py:994:print_rank_0] [(11, 0, 154533888)]`

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 1]
```diff
@@ -531,6 +543,50 @@ def __init__(self,
         self._enable_universal_checkpoint()
         self._param_slice_mappings = self._create_param_mapping()
 
+    def get_partition_info_from_bwd_params(self):
+        params_to_parition = {}
```
Comment (tjruwase) [inline deepspeed/runtime/zero/stage_1_and_2.py:547]: I don't see the return of `params_to_parition`.

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 2]
```diff
@@ -708,6 +764,8 @@ def initialize_gradient_partitioning_data_structures(self):
                 self.total_grads_in_partition[i][partition_id] = 0
                 self.initialize_gradient_partition(i, param_group, partition_id)
                 self.is_partition_reduced[i][partition_id] = False
+
+            for partition_id in range(total_partitions):
```
Comment (tjruwase) [inline deepspeed/runtime/zero/stage_1_and_2.py:768]: Why is this needed? It seems not to change the logic.

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 3]
```diff
@@ -1630,6 +1719,9 @@ def get_flat_partition(self,
             # we need to offset to get to the right element
             if i == 0 and first_offset > 0:
                 tensor_offset = first_offset
+                if tensor_offset > num_elements:
```
Comment (tjruwase) [inline deepspeed/runtime/zero/stage_1_and_2.py:1722]: I have two questions here:
1. How can `tensor_offset > num_elements` be True? Does this mean the first tensor is skipped entirely from the flattened result?
2. Line 1723 seems to be a noop because of 1717?

File: deepspeed/runtime/zero/stage_1_and_2.py (hunk 1 above)
Comment (Liangliang-Ma) [reply deepspeed/runtime/zero/stage_1_and_2.py]: Sorry for the miss.

File: deepspeed/runtime/zero/stage_1_and_2.py (hunk 2 above)
Comment (Liangliang-Ma) [reply deepspeed/runtime/zero/stage_1_and_2.py:769]: The next line under this (R771) calls `self.get_first_param_index`. This code in this function look up the `self.param_to_partition_ids[group_id][param_id]` (R743) in the order of param_id order, which will cause error. Because `self.param_to_partition_ids` is set in `def initialize_gradient_partitioning_data_structures`(R747) in the loop of `param_group` and `total_partitions`. For example, the first param in the first group maybe not set to first partition. So the dict `self.param_to_partition_ids[0]` will have no key of "0" when finish the first loop of `total_partition `of the first group(L704). After this change, all of  `self.param_to_partition_ids[group_id][param_id]` exists, which will avoid no key error.

Comment (tjruwase) [reply deepspeed/runtime/zero/stage_1_and_2.py:769]: But I think `self.param_to_partition_ids[*][*]` is set by `self.initialize_gradient_partition(...)` in (R766), no? 

Or is it because you are changing the order of `self.param_to_parition_ids` initialization, so we need to wait for the entire dict to be initialized before any param_id can be safely read?

File: deepspeed/runtime/zero/stage_1_and_2.py (hunk 3 above)
Comment (Liangliang-Ma) [reply deepspeed/runtime/zero/stage_1_and_2.py:1723]: This is because the first_offset maybe not corresponding to the first tensor of the group. So ` tensor_offset > num_elements`  can be True. I found that this would set every `tensor.grad ` a wrong position, although the total tensor size is right. I will fix it asap.

Comment (tjruwase) [reply deepspeed/runtime/zero/stage_1_and_2.py:1723]: Two more questions:
1. So, it is possible for the first tensor to be missing from the flattened version of the partition. Is this correct? Previously, every tensor (sometimes a subset of the first tensor) was included in the flattened version. 

2. Also, is it only the first tensor that could be missing from the flattened version, or could other tensors be missing?

File: deepspeed/runtime/zero/stage_1_and_2.py (hunk 2 above)
Comment (Liangliang-Ma) [reply deepspeed/runtime/zero/stage_1_and_2.py:771]: > But I think `self.param_to_partition_ids[*][*]` is set by `self.initialize_gradient_partition(...)` in (R766), no?
Yes, exactly. It's set by `self.initialize_gradient_partition(...)`.

> Or is it because you are changing the order of `self.param_to_parition_ids` initialization, so we need to wait for the entire dict to be initialized before any param_id can be safely read?
Yes, you are right:)

File: deepspeed/runtime/zero/stage_1_and_2.py (hunk 3 above)
Comment (Liangliang-Ma) [reply deepspeed/runtime/zero/stage_1_and_2.py:1725]: No. Every tensor should be included in the flattened version I think.

Comment (tjruwase) [reply deepspeed/runtime/zero/stage_1_and_2.py:1725]: Sorry, I still don't get it :). Line 1724 will exclude the first tensor from the flattened version, no?

Comment (Liangliang-Ma) [reply deepspeed/runtime/zero/stage_1_and_2.py:1725]: The design on Line 1724 will be deprecated. I am working on a new design :).

Comment (tjruwase) [reply deepspeed/runtime/zero/stage_1_and_2.py:1725]: Got it. Thanks so much!

**Comment by tjruwase** [comment]
@Liangliang-Ma, just wanted to check on this? Is there anything we can do to help? Thanks!

**Comment by Liangliang-Ma** [comment]
@tjruwase any comments? thanks!

**Comment by tjruwase** [comment]
@Liangliang-Ma, thanks for the update. I will review asap. In the meantime, can you please respond to the unresolved conversations, I suspect they are now addressed.

File: deepspeed/runtime/zero/stage_1_and_2.py (hunk 1 above)
Comment (Liangliang-Ma) [reply deepspeed/runtime/zero/stage_1_and_2.py]: solved. thanks

**Comment by Liangliang-Ma** [comment]
> @Liangliang-Ma, thanks for the update. I will review asap. In the meantime, can you please respond to the unresolved conversations, I suspect they are now addressed.

Sure. Previous issues that you mentioned are all solved or outdated. I marked them resolved. Thanks!

**Comment by tjruwase** [comment]
@Liangliang-Ma, apologies that we never merged this PR. I think the optimization is really good and would like to merge it. Do you have some bandwidth to refresh this? Thanks!
